### PR TITLE
This is the Pull Request for the second Project in the Ruby modules

### PR DESCRIPTION
**Pull request by jose-Abel**
Hello, 

We are presenting our code for the second project in Ruby's module where we got requested to create our own versions of some of the methods from the Enumerable class, for the methods: my_each, my_each_with_index, my_select, my_all? my_any? my_none? my_count, my_map, and my_inject. Also modifying the map method to take either a block or a proc.

Hoping to get amazing feedback! happy coding!

### CODE REVIEW

**Comment by Dandush03** [comment]
# Status: :red_circle: Required Changes :red_circle: 

## Good job for the first try, this is a hard assignment. There are some changes you must perform in order to make a new submit. :nerd_face: 

### I will point out all the errors and then dive into your code and see what might be the issue in most cases.
> ### NOTICE: I Won't Point Out The Same Error Twice

### #my_each
- [ ] returns an Enumerator if no block is given
> remember that all enumerators when no block is given should return an enumerator. check out [this doc](https://ruby-doc.org/core-2.6.4/Array.html#method-i-each)
### #my_each_with_index
- [ ] returns an enumerator if no block is given
- [ ] when a block is given when self is an array, returns the array itself after calling the given block once for each element in self
- [ ] when a block is given when self is an array behaves like #each_with_index calls the given block once for each element in self
- [ ] when a block is given when self is a range returns the range itself after calling the given block once for each element in self
> checkout [this doc](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-each_with_index) to get a better idea of the behavior of `my_each_with_index`
### #my_select
- [ ] returns an enumerator if no block is given
> same as before, an enumerator return enumerator if no block is given
### #my_all?
- [ ] when no block or argument is given returns true when none of the collection members are false or nil
- [ ] when no block or argument is given returns false when one of the collection members are false or nil
- [ ] when a class is passed as an argument returns true if all of the collection is a member of such class::Integer
- [ ] when a class is passed as an argument returns true if all of the collection is a member of such class::Numeric
- [ ] when a class is passed as an argument returns false if any of the collection is not a member of such class
- [ ] when a Regex is passed as an argument returns true if all of the collection matches the Regex
- [ ] when a Regex is passed as an argument returns false if any of the collection does not match the Regex
- [ ] when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern
- [ ] when a pattern other than Regex or a Class is given returns false if any of the collection does not match the pattern
> Check out the following documentation since this error is being repeated in #my_none? & #my_any?
> [#my_all docs]()
> [#my_all docs]()
> [#my_all docs]()

**Review by Dandush03 — CHANGES_REQUESTED** [review]
# Status: :red_circle: Required Changes :red_circle: 

## Good job for the first try, this is a hard assignment. There are some changes you must perform in order to make a new submit. :nerd_face: 

### I will point out all the errors and then dive into your code and see what might be the issue in most cases.
> ### NOTICE: I Won't Point Out The Same Error Twice

### #my_each
- [x] returns an Enumerator if no block is given
> remember that all enumerators when no block is given should return an enumerator. check out [this doc](https://ruby-doc.org/core-2.6.4/Array.html#method-i-each)
### #my_each_with_index
- [x] returns an enumerator if no block is given
- [x] when a block is given when self is an array, returns the array itself after calling the given block once for each element in self
- [x] when a block is given when self is an array behaves like #each_with_index calls the given block once for each element in self
- [x] when a block is given when self is a range returns the range itself after calling the given block once for each element in self
> checkout [this doc](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-each_with_index) to get a better idea of the behavior of `my_each_with_index`
### #my_select
- [x] returns an enumerator if no block is given
> same as before, an enumerator return enumerator if no block is given
### #my_all?
- [x] when no block or argument is given returns true when none of the collection members are false or nil
- [x] when no block or argument is given returns false when one of the collection members are false or nil
- [x] when a class is passed as an argument returns true if all of the collection is a member of such class::Integer
- [x] when a class is passed as an argument returns true if all of the collection is a member of such class::Numeric
- [x] when a class is passed as an argument returns false if any of the collection is not a member of such class
- [x] when a Regex is passed as an argument returns true if all of the collection matches the Regex
- [x] when a Regex is passed as an argument returns false if any of the collection does not match the Regex
- [x] when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern
- [x] when a pattern other than Regex or a Class is given returns false if any of the collection does not match the pattern
> Check out the following documentation since this error is being repeated in #my_none? & #my_any?
> [#all? docs](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-all-3F)
> [#any? docs](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-any-3F)
> [#none? docs](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-none-3F)
### #my_count 
- [x] returns the number of items in enum through enumeration::range
- [x] counts the number of items in enum that are equal to the item if an argument is given
> Take a look to [this documentation](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-count) if you have any issue
### #my_map 
- [x] returns an Enumerator if no block is given
- [x] executes only the proc when both a block and a proc are given and returns a new array with the results of running the proc once for every element in enum.
> Take a look to [this documentation](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-map) if you have any issue
### #my_inject 
- [x] when a block is given without an initial value combines all elements of enum by applying a binary operation, specified by a block::range
- [x] when a block is given with an initial value combines all elements of enum by applying a binary operation, specified by a block
- [x] when a symbol is specified without an initial value combines each element of the collection by applying the symbol as a named method
- [x] when a symbol is specified without an initial value combines each element of the collection by applying the symbol as a named method:range
- [x] when a symbol is specified with an initial value combines each element of the collection by applying the symbol as a named method
- [x] when a symbol is specified with an initial value combines each element of the collection by applying the symbol as a named method::range
> You can find more info regarding #inject behavior in [this doc](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject)

# If you need any help or you find yourself in trouble to fix one of this issue, don't hesitate to contact me through slack

# Happy Coding :nerd_face: :computer:

**Review by kitpao — CHANGES_REQUESTED** [review]
# 👋 Hello @jose-Abel and @t-yanick 

👍 Good job fixing some of the issues in your enumerables, there are some issues remaining, kindly check them below 🔽 


### Methods:

#### #my_each_with_index
 - [x] `[8,8].my_each_with_index {|x, i| puts "#{x}, #{i}"}` should return 8, 0 /n 8, 1. But yours return 8, 0 /n 8, 0. **Hint:** Remember that #find_index() only finds the first item that matches, that is why it is returning always the 0 index.

#### #my_all?
 - [x] `[1, false].my_all?` should be false because not all of the elements are true. Your method returns true

#### #my_any?
- [x] `[false, false].my_any?` should be false because not even one of the elements is true. Your method returns true
- [x] `[1, 2].my_any?(String)` should be false because there is not a string in the array. Your method returns always true (even with Regexp or Integers)
- [x] `["bird", "dog"].my_any?('cat')` should be false since there is no cat. Your method returns always true

#### #my_none?
- [x] All these should return true, but they return false:
  ```
  [false, false].my_none?
  [1, 2].my_none?(String)
  (1..3).my_none?{|x| x > 4}
  ```
#### #my_count
- [x] `(1..3).my_count{|x| x >= 2}` should return 2 because there are 2 elements that have that condition in the block. Your method returns nil
#### #my_map
- [x] Your code doesn't know what to do when both a proc and a block are given. It should give priority to the proc
image

#### #my_inject
- [x] If there is no block or argument, your method should return a local jump error ([how to raise errors here](https://stackoverflow.com/questions/4325507/how-to-force-a-localjumperror-on-return-in-ruby))
- [x] Your method should be able to manage symbols for operations (:+, :*, etc) [Check the examples in the documentation here](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject)

### ♻️ Status: Your project requires some additional changes ♻️
➡️ Please request another review once you are ready 📝 
___
##### 👉 If you have any doubts you can contact me in slack as @kitpao

🔗 __[Github](https://github.com/kitpao) | [Twitter](https://twitter.com/Kitpao1) | [LinkedIn](https://www.linkedin.com/in/kitzia-paola-vidal/) | [Medium](https://medium.com/@kitpao91)__

File: enumerable.rb [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+
+# rubocop:disable Style/IdenticalConditionalBranches
+
+# rubocop:disable Style/NilComparison
+
+# rubocop:disable Metrics/ModuleLength
+
+# rubocop:disable Style/RedundantReturn
+
+# rubocop:disable Style/IfUnlessModifier
```
Comment (kitpao) [inline enumerable.rb]: - [x] Sorry for not mentioning it before, but the only lintern you can skip is the Style/CaseEquality. For the others, you should refactor your code to follow the standard good coding practices
image

**Review by kitpao — CHANGES_REQUESTED** [review]
# 👋 Hi again!

🥳 Good job, now you only have 3 issues to address,

Please, don't forget to check the inline comment at the end 🔽 it's about the linters

- [x] Your **#my_none** should take into account if the condition specified in the block is true or false for the argument, `(1..3).none?{|x| x > 2}` should be false (because there is a number greater than 2) and your program is returning true
- [x] **#my_count** should also work without a block, `(1..3).my_count` should return 3. **Hint:** `.size` works with ranges :)
- [x] **#my_inject** should also allow multiplications, `(2..3).my_inject(20, :*)` should be 120 (20 * 2 * 3), your program is returning 26 (20 + 2*3 )

### ♻️ Status: Your project requires some additional changes ♻️
➡️ Please request another review once you are ready 📝 
___
##### 👉 If you have any doubts you can contact me in slack as @kitpao

🔗 __[Github](https://github.com/kitpao) | [Twitter](https://twitter.com/Kitpao1) | [LinkedIn](https://www.linkedin.com/in/kitzia-paola-vidal/) | [Medium](https://medium.com/@kitpao91)__

**Review by MarioDena — APPROVED** [review]
Hello Team! Great work implementing the requested changes. 👍 

The project is approved. However I would like to make a general recommendation, Either write more descriptive commits or learn how to rebase using git. Repetition of messages like "fixing rubocops errors" makes your repositories look unprofessional. This is a requirement for all projects (Found under general requirements ) but since the previous reviews failed to point it out, I'll leave a note here about it. Please keep that in mind for future projects. 

Cheers! 🎖️

**Pull request merged**